extensions/azure/storage/AzureDataLakeStorage.cpp

```diff
@@ -61,4 +61,29 @@ bool AzureDataLakeStorage::deleteFile(const storage::DeleteAzureDataLakeStorageP
   }
 }
 
+std::optional<uint64_t> AzureDataLakeStorage::fetchFile(const FetchAzureDataLakeStorageParameters& params, io::BaseStream& stream) {
+  try {
+    auto fetch_res = data_lake_storage_client_->fetchFile(params);
+
+    std::array<uint8_t, 4096> buffer;
+    size_t write_size = 0;
+    if (fetch_res.FileSize < 0) return 0;
+    while (write_size < gsl::narrow<uint64_t>(fetch_res.FileSize)) {
+      const auto next_write_size = (std::min)(gsl::narrow<size_t>(fetch_res.FileSize) - write_size, buffer.size());
+      if (!fetch_res.Body->Read(buffer.data(), gsl::narrow<std::streamsize>(next_write_size))) {
+        return -1;
+      }
+      const auto ret = stream.write(buffer.data(), next_write_size);
+      if (io::isError(ret)) {
+        return -1;
+      }
+      write_size += next_write_size;
+    }
```

Comment:
Not sure if it's worth the extra effort, but if you could wrap the download result in an `InputStream`, then you could reuse `internal::pipe` (which has similar code to this function) or just return the stream directly.]
It would also probably be easier to mock `fetchFile`.

Reply:
Good idea, saves us some code duplication, updated in 714793593b09ed8b33ac7db2547c74258467e748

Reply:
I think it would be better if `fetchFile` itself returned an `InputStream`. This way mocking it would be as easy as returning some data in a `BufferStream`.

Reply:
I see your point, updated in 0f01db34bb1cc4576c7ed709dd1b9ef71801b43d